src/ccpp_constituent_prop_mod.F90

```diff
@@ -401,441 +560,670 @@ subroutine ccp_set_const_index(this, index, errcode, errmsg)
+
+   subroutine ccp_default_value(this, val_out, errcode, errmsg)
+
+      ! Dummy arguments
+      class(ccpp_constituent_properties_t), intent(in)  :: this
+      real(kind_phys),                      intent(out) :: val_out
+      integer,                              intent(out) :: errcode
+      character(len=*),                     intent(out) :: errmsg
+
+      if (this%is_instantiated(errcode, errmsg)) then
+         val_out = this%const_default_value
+      else
+         val_out = kphys_unassigned
+      end if
+
+   end subroutine ccp_default_value
+
+   !########################################################################
+
+   subroutine ccp_has_default(this, val_out, errcode, errmsg)
+
+      ! Dummy arguments
+      class(ccpp_constituent_properties_t), intent(in)  :: this
+      logical,                              intent(out) :: val_out
+      integer,                              intent(out) :: errcode
+      character(len=*),                     intent(out) :: errmsg
+      ! Local variable
+      character(len=*), parameter :: subname = 'ccp_has_default'
+
+      if (this%is_instantiated(errcode, errmsg)) then
+         val_out = this%const_default_value /= kphys_unassigned
+      else
+         val_out = .false.
+      end if
+
+   end subroutine ccp_has_default
+
+   !########################################################################
+   !
+   ! CCPP_MODEL_CONSTITUENTS_T (constituent field data) methods
+   !
+   !########################################################################
+
+   logical function ccp_model_const_locked(this, errcode, errmsg, warn_func)
+      ! Return .true. iff <this> is locked (i.e., ready to use)
+      ! Optionally fill out <errcode> and <errmsg> if object not initialized
+
+      ! Dummy arguments
+      class(ccpp_model_constituents_t), intent(in)  :: this
+      integer,          optional,       intent(out) :: errcode
+      character(len=*), optional,       intent(out) :: errmsg
+      character(len=*), optional,       intent(in)  :: warn_func
+      ! Local variable
+      character(len=*), parameter :: subname = 'ccp_model_const_locked'
+
+      call initialize_errvars(errcode, errmsg)
+      ccp_model_const_locked = .false.
+      ! Use an initialized hash table as double check
+      if (this%hash_table%is_initialized()) then
+         ccp_model_const_locked = this%table_locked .and. this%data_locked
+         if ( (.not. (this%table_locked .and. this%data_locked)) .and.                                 &
+              present(errmsg) .and. present(warn_func)) then
+            ! Write a warning as a courtesy to calling function but do not set
+            !   errcode (let caller decide).
+            write(errmsg, *) trim(warn_func),                                 &
+                 ' WARNING: Model constituents not ready to use'
+         end if
+      else
+         if (present(warn_func)) then
```

Comment:
I like the change from `set_errvars` to `append_errvars`, however, this pattern shows up so often, I wonder if `append_errvars` could take two new arguments:
`subname` would always be passed and be the backup.
`caller` would be the optional but preferred item.
Given that `append_errvars` is called (I think) 57 times in this module, it would really clean up this module.

Reply:
I like this idea, but am unclear on the difference between "subname" and "caller"? Would caller be the name of the function that produced the error where subname is the name of the subroutine that append_errvars is being called from?

Reply:
See my suggested call change at line 904 (just added). Does that help?

Reply:
yes! thank you!

Reply:
Yow, lots of work but it does look a lot better (IMHO at least).